Source/Core/VideoCommon/TmemBase.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <array>
+
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/TmemBase.h"
+
+namespace TmemBase
+{
+// On TMEM configuration changed:
+// 1. invalidate stage.
+
+struct texture_unit_state
+{
+  enum State
+  {
+    INVALID, // Configuration has changed
+    VALID, // But not cached due to either being too big, or overlapping with another valid texture unit
+    CACHED,
+  };
+
+  TexImage1 even;
+  TexImage2 odd;
+  State state;
+
+  int even_base;
+  int even_size;
+  int odd_base;
+  int odd_size;
+};
+static std::array<texture_unit_state, 8> s_unit;
```

Comment:
Is there really no other place to place this (like in the actual cache)? This introduces more file-scope/global mutable state.

Reply:
I feel really uneasy about how complicated texture cache is getting and just how complicated it is to understand.

This part is still WIP, this state will need to be stored somewhere, and it will need initialized to zero, and something will need to be done on DoState (either we invalidate the whole thing, or for correctness we save anything that's tagged as "cached")